File: C++/016. 3Sum Closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int res=nums[0]+nums[1]+nums[2]-target;
        int diff=abs(nums[0]+nums[1]+nums[2]-target);
        int size = nums.size();
        int left,right;
        sort(nums.begin(),nums.end());
        for (int i=0;i<size-2;i++)
        {
            left = i+1;
            right = size-1;
            while(left<right)
            {
                diff = nums[i]+nums[left]+nums[right]-target;
                if (abs(diff)<abs(res))
                {
                    res = diff;
                }
                if(diff<0){left++;}
                else if(diff>0){right--;}
                else
                {
                    return target;
                }
            }
            while(nums[i+1]==nums[i]){i++;}
        }
        res += target;
        return res;
    }
};

File: python/016. 3Sum Closest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        n = len(nums)
        nums.sort()
        left, right = 0, n-1
        d = float('inf')
        res = float('inf')
        for i in range(n-2):
            left, right = i + 1, n-1
            while left < right:
                val = nums[left] + nums[right] + nums[i]
                if abs(val - target) < d: 
                    res = val
                    d = abs(val - target)
                if val == target:
                    break
                elif val < target:
                    left += 1
                else:
                    right -= 1
        return res